flask_jwt_extended/config.py

```diff
@@ -9,12 +11,37 @@
 from typing import Union
 
 from flask import current_app
-from flask.json import JSONEncoder
+from flask.json.provider import _default
 from jwt.algorithms import requires_cryptography
 
 from flask_jwt_extended.typing import ExpiresDelta
 
 
+class JSONEncoder(json.JSONEncoder):
+    """The default JSON encoder. Handles extra types compared to the
+    built-in :class:`json.JSONEncoder`.
+
+    -   :class:`datetime.datetime` and :class:`datetime.date` are
+        serialized to :rfc:`822` strings. This is the same as the HTTP
+        date format.
+    -   :class:`decimal.Decimal` is serialized to a string.
+    -   :class:`uuid.UUID` is serialized to a string.
+    -   :class:`dataclasses.dataclass` is passed to
+        :func:`dataclasses.asdict`.
+    -   :class:`~markupsafe.Markup` (or any object with a ``__html__``
+        method) will call the ``__html__`` method to get a string.
+
+    """
+
+    def default(self, o: Any) -> Any:
+        """Convert ``o`` to a JSON serializable type. See
+        :meth:`json.JSONEncoder.default`. Python does not support
+        overriding how basic types like ``str`` or ``list`` are
+        serialized, they are handled before this method.
+        """
+        return _default(o)
```

Comment:
Hrm... the thing I am sold on is that the only reason we have this `config.json_encoder` property is so that we can use the same encoder for the flask app as for pyjwt (per a8605502).

Even though this `_default` version does more then standard `json.JSONEncoder`, we could still run into situations where someone is using a custom thing that has additional features, and breaks when trying to use those custom things with a JWT.

I want to dig into the internals here a bit more and see if there is a way to bridge this gap. But if that isn't going to be easy or I don't have enough time to look, I'm happy to go with this approach initially as a stop gap solution!

👍

Reply:
The solution using `_default` is not optimal, as you said it might break some setups.

Reply:
Actually in Flask >= 2.2 there is the `json_provider_class` property on the `current_app`. I think this would be the way to go to support other JSON Providers. We just need to find a way to make this Flask <2.2 / >= 2.2 compatible... If this is even a goal?

Reply:
Would something as simple as having an `except AttributeError:` and using the old `current_app.json_encoder` in that case to support both versions? If so I would just as soon do that, but I'm not opposed to updating the `setup.py` to have the newer release of flask be the minimum supported version if it isn't so simple.

Reply:
Sadly it's not as simple as that. The base flask JSON provider class (https://github.com/pallets/flask/blob/main/src/flask/json/provider.py#L20) does not require that custom implementations provide a `default` method, so it cannot be expected that this method exists.